File: ubel/cvss_parser.py
"""
cvss_parser.py — CVSS parser supporting v2, v3.0, v3.1, v4.0, and Ubuntu severity labels.

Pure-Python port of cvss_parser.js.
Zero external dependencies (uses cvss40.py for v4 scoring).

API
---
    parse(vector)                  -> {"score": str|None, "severity": str}
    process_vulnerability(vuln)    -> mutates the OSV vuln dict in-place
"""

from __future__ import annotations

import math
import re
from typing import Dict, Optional

from .cvssv40 import CVSS40


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

_UBUNTU_LABELS = {"critical", "high", "medium", "low", "negligible", "untriaged"}


def _score_to_severity(score_str: str) -> str:
    try:
        score = float(score_str)
    except (ValueError, TypeError):
        return "unknown"
    if math.isnan(score) or score == 0.0:
        return "unknown"
    if score < 4.0:  return "low"
    if score < 7.0:  return "medium"
    if score < 9.0:  return "high"
    return "critical"


def _parse_metrics(vector: str) -> Dict[str, str]:
    """Strip the version prefix and return a key→value dict."""
    body = re.sub(r"^CVSS:\d+\.\d+/", "", vector).strip("()")
    result: Dict[str, str] = {}
    for part in body.split("/"):
        if ":" in part:
            k, v = part.split(":", 1)
            result[k] = v
    return result


# ---------------------------------------------------------------------------
# CVSSv4 scoring
# ---------------------------------------------------------------------------

def _extract_cvss4_score(vector: str) -> Optional[str]:
    try:
        cvss = CVSS40(vector)
        return f"{cvss.score:.1f}"
    except Exception:
        return None


# ---------------------------------------------------------------------------
# CVSSv3.x base-score calculation
# ---------------------------------------------------------------------------

def _round_up(value: float) -> float:
    """CVSS v3 'always round up' to 1 decimal place."""
    int_val = round(value * 100000)
    if int_val % 10000 == 0:
        return int_val / 100000
    return (math.floor(int_val / 10000) + 1) / 10


def _extract_cvss3_score(vector: str) -> Optional[str]:
    try:
        m = _parse_metrics(vector)

        AV_W  = {"N": 0.85, "A": 0.62, "L": 0.55, "P": 0.20}
        AC_W  = {"L": 0.77, "H": 0.44}
        PR_W  = {"N": 0.85, "L": 0.62, "H": 0.27}        # scope unchanged
        PR_SC = {"N": 0.85, "L": 0.68, "H": 0.50}        # scope changed
        UI_W  = {"N": 0.85, "R": 0.62}
        CIA_W = {"N": 0.00, "L": 0.22, "H": 0.56}

        S  = m.get("S", "U")
        AV = AV_W.get(m.get("AV", "N"), 0.85)
        AC = AC_W.get(m.get("AC", "L"), 0.77)
        PR = (PR_SC if S == "C" else PR_W).get(m.get("PR", "N"), 0.85)
        UI = UI_W.get(m.get("UI", "N"), 0.85)
        C  = CIA_W.get(m.get("C", "N"), 0.00)
        I  = CIA_W.get(m.get("I", "N"), 0.00)
        A  = CIA_W.get(m.get("A", "N"), 0.00)

        ISS = 1 - (1 - C) * (1 - I) * (1 - A)
        if S == "U":
            IS = 6.42 * ISS
        else:
            IS = 7.52 * (ISS - 0.029) - 3.25 * ((ISS - 0.02) ** 15)

        if IS <= 0:
            return "0.0"

        exploitability = 8.22 * AV * AC * PR * UI

        if S == "U":
            score = min(IS + exploitability, 10.0)
        else:
            score = min(1.08 * (IS + exploitability), 10.0)

        return f"{_round_up(score):.1f}"
    except Exception:
        return None


# ---------------------------------------------------------------------------
# CVSSv2 base-score calculation
# ---------------------------------------------------------------------------

def _extract_cvss2_score(vector: str) -> Optional[str]:
    try:
        m = _parse_metrics(vector)

        AV_W  = {"N": 1.0,   "A": 0.646, "L": 0.395}
        AC_W  = {"L": 0.71,  "M": 0.61,  "H": 0.35}
        Au_W  = {"N": 0.704, "S": 0.56,  "M": 0.45}
        CIA_W = {"N": 0.0,   "P": 0.275, "C": 0.660}

        AV = AV_W.get(m.get("AV", "N"), 1.0)
        AC = AC_W.get(m.get("AC", "L"), 0.71)
        Au = Au_W.get(m.get("Au", "N"), 0.704)
        C  = CIA_W.get(m.get("C",  "N"), 0.0)
        I  = CIA_W.get(m.get("I",  "N"), 0.0)
        A  = CIA_W.get(m.get("A",  "N"), 0.0)

        f_impact = 0 if (C == 0 and I == 0 and A == 0) else 1.176
        impact   = 10.41 * (1 - (1 - C) * (1 - I) * (1 - A))
        exploit  = 20 * AV * AC * Au

        base = ((0.6 * impact) + (0.4 * exploit) - 1.5) * f_impact
        return str(round(base * 10) / 10)
    except Exception:
        return None


# ---------------------------------------------------------------------------
# Main parse()
# ---------------------------------------------------------------------------

def parse(vector: object) -> Dict[str, Optional[str]]:
    """
    Parse a CVSS vector string (or Ubuntu severity label).

    Returns:
        {"score": str|None, "severity": str}
    """
    if not vector or not isinstance(vector, str):
        return {"score": None, "severity": "unknown"}

    v = vector.strip()

    # Ubuntu / plain severity label
    if v.lower() in _UBUNTU_LABELS:
        sev = v.lower()
        if sev == "negligible":
            sev = "low"
        if sev == "untriaged":
            sev = "unknown"
        return {"score": None, "severity": sev}

    # Plain numeric score
    if re.fullmatch(r"\d+(\.\d+)?", v):
        return {"score": v, "severity": _score_to_severity(v)}

    # CVSSv4
    if v.startswith("CVSS:4."):
        score = _extract_cvss4_score(v)
        severity = _score_to_severity(score) if score is not None else "unknown"
        return {"score": score, "severity": severity}

    # CVSSv3.x
    if v.startswith("CVSS:3."):
        score = _extract_cvss3_score(v)
        severity = _score_to_severity(score) if score is not None else "unknown"
        return {"score": score, "severity": severity}

    # CVSSv2 (no version prefix)
    if v.startswith("AV:") or v.startswith("(AV:"):
        score = _extract_cvss2_score(v)
        severity = _score_to_severity(score) if score is not None else "unknown"
        return {"score": score, "severity": severity}

    return {"score": None, "severity": "unknown"}


# ---------------------------------------------------------------------------
# process_vulnerability()  — mirrors Python process_vulnerability behaviour
# ---------------------------------------------------------------------------

def process_vulnerability(vuln: dict) -> None:
    """
    Mutate an OSV vulnerability dict in-place, setting:
        vuln["severity"]        str
        vuln["severity_score"]  float | None
        vuln["severity_vector"] str | None
    """
    severity_list = vuln.get("severity")

    if isinstance(severity_list, list) and severity_list:
        first  = severity_list[0]
        vector = first.get("score", "")
        type_  = (first.get("type") or "").lower()

        if type_ in ("ubuntu_cve", "ubuntu"):
            vuln["severity"]        = vector
            vuln["severity_score"]  = None
            vuln["severity_vector"] = None
            return

        result   = parse(vector)
        score    = result["score"]
        severity = result["severity"]

        vuln["severity"]        = severity
        vuln["severity_score"]  = float(score) if score is not None else None
        vuln["severity_vector"] = vector

        # Fallback bucket if severity still unknown but score exists
        if severity in ("unknown", None) and score is not None:
            vuln["severity"] = _score_to_severity(score)
        return

    vuln["severity"]        = "unknown"
    vuln["severity_score"]  = None
    vuln["severity_vector"] = None

File: node/sca/cvss_parser.js
/**
 * CVSS Parser — supports v2, v3.0, v3.1, v4.0, and Ubuntu string severity labels.
 *
 * The upstream OSV "severity" array looks like:
 *   [ { "type": "CVSS_V3", "score": "CVSS:3.1/AV:N/…" }, … ]
 *   [ { "type": "UBUNTU_CVE",  "score": "medium" }, … ]
 *
 * parse(vector) → { score: string|null, severity: string }
 * processVulnerability(vuln)  mutates the OSV vuln object in-place,
 *   exactly mirroring the Python CVSS_Parser.process_vulnerability behaviour.
 */

import { CVSS40 } from "./cvss40.js";

// ── Score → severity bucket (mirrors Python fallback logic) ──────────────────
function scoreToSeverity(scoreStr) {
  const score = parseFloat(scoreStr);
  if (isNaN(score)) return "unknown";
  if (score === 0.0) return "unknown";
  if (score < 4.0) return "low";
  if (score < 7.0) return "medium";
  if (score < 9.0) return "high";
  return "critical";
}

// ── Numeric extraction helpers ────────────────────────────────────────────────

/**
 * Extract the base score from a CVSS vector string.
 * We compute it from the mandatory metrics rather than relying on a heavy
 * library, because the 'cvss' npm package API differs across versions.
 *
 * For our purposes (severity bucketing) a look-up of the pre-calculated
 * score embedded in the vector itself is sufficient when present, otherwise
 * we fall back to parsing the vector metrics.
 *
 * OSV frequently embeds the score directly as a numeric string instead of
 * a full vector — we handle that too.
 */

/**
 * Try to parse a CVSS vector and return { score, severity }.
 * Handles:
 *   CVSS:4.0/…   → CVSSv4
 *   CVSS:3.1/…   → CVSSv3.1
 *   CVSS:3.0/…   → CVSSv3.0
 *   AV:…         → CVSSv2 (no prefix)
 *   plain number → treat as pre-computed score
 *   Ubuntu label → pass through as severity, no numeric score
 */
export function parse(vector) {
  if (!vector || typeof vector !== "string") return { score: null, severity: "unknown" };

  const v = vector.trim();

  // ── Ubuntu / plain severity labels ────────────────────────────────────────
  const UBUNTU_LABELS = new Set(["critical", "high", "medium", "low", "negligible", "untriaged"]);
  if (UBUNTU_LABELS.has(v.toLowerCase())) {
    const sev = v.toLowerCase() === "negligible" ? "low" : v.toLowerCase();
    return { score: null, severity: sev === "untriaged" ? "unknown" : sev };
  }

  // ── Plain numeric score (some OSV entries use this) ───────────────────────
  if (/^\d+(\.\d+)?$/.test(v)) {
    const sev = scoreToSeverity(v);
    return { score: v, severity: sev };
  }

  // ── CVSSv4 ────────────────────────────────────────────────────────────────
  if (v.startsWith("CVSS:4.")) {
    const score = extractCVSS4Score(v);
    const severity = score !== null ? scoreToSeverity(score) : "unknown";
    return { score, severity };
  }

  // ── CVSSv3.x ─────────────────────────────────────────────────────────────
  if (v.startsWith("CVSS:3.")) {
    const score = extractCVSS3Score(v);
    const severity = score !== null ? scoreToSeverity(score) : "unknown";
    return { score, severity };
  }

  // ── CVSSv2 (no version prefix) ────────────────────────────────────────────
  if (v.startsWith("AV:") || v.startsWith("(AV:")) {
    const score = extractCVSS2Score(v);
    const severity = score !== null ? scoreToSeverity(score) : "unknown";
    return { score, severity };
  }

  return { score: null, severity: "unknown" };
}

function extractCVSS4Score(vector) {
  try {
    const cvss = new CVSS40(vector);
    return cvss.score.toFixed(1);
  } catch {
    return null;
  }
}

// ── CVSSv3.x base-score calculation ───────────────────────────────────────────
function extractCVSS3Score(vector) {
  try {
    const metrics = parseMetrics(vector);

    const AV_W  = { N: 0.85, A: 0.62, L: 0.55, P: 0.20 };
    const AC_W  = { L: 0.77, H: 0.44 };
    const PR_W  = {
      // Scope unchanged
      N: 0.85, L: 0.62, H: 0.27,
    };
    const PR_SC = { N: 0.85, L: 0.68, H: 0.50 }; // scope changed
    const UI_W  = { N: 0.85, R: 0.62 };
    const CIA_W = { N: 0.00, L: 0.22, H: 0.56 };

    const S   = metrics["S"]  || "U";
    const AV  = AV_W[metrics["AV"]]  ?? 0.85;
    const AC  = AC_W[metrics["AC"]]  ?? 0.77;
    const prTable = S === "C" ? PR_SC : PR_W;
    const PR  = prTable[metrics["PR"]] ?? 0.85;
    const UI  = UI_W[metrics["UI"]]  ?? 0.85;
    const C   = CIA_W[metrics["C"]]  ?? 0.00;
    const I   = CIA_W[metrics["I"]]  ?? 0.00;
    const A   = CIA_W[metrics["A"]]  ?? 0.00;

    const ISS  = 1 - (1 - C) * (1 - I) * (1 - A);
    const IS   = S === "U"
      ? 6.42 * ISS
      : 7.52 * (ISS - 0.029) - 3.25 * Math.pow(ISS - 0.02, 15);

    if (IS <= 0) return "0.0";

    const exploitability = 8.22 * AV * AC * PR * UI;

    let score;
    if (S === "U") {
      score = Math.min(IS + exploitability, 10);
    } else {
      score = Math.min(1.08 * (IS + exploitability), 10);
    }

    return roundUp(score).toFixed(1);
  } catch {
    return null;
  }
}

// ── CVSSv2 base-score calculation ─────────────────────────────────────────────
function extractCVSS2Score(vector) {
  try {
    const metrics = parseMetrics(vector);

    const AV_W = { N: 1.0, A: 0.646, L: 0.395 };
    const AC_W = { L: 0.71, M: 0.61, H: 0.35 };
    const Au_W = { N: 0.704, S: 0.56, M: 0.45 };
    const CIA_W = { N: 0.0, P: 0.275, C: 0.660 };

    const AV = AV_W[metrics["AV"]] ?? 1.0;
    const AC = AC_W[metrics["AC"]] ?? 0.71;
    const Au = Au_W[metrics["Au"]] ?? 0.704;
    const C  = CIA_W[metrics["C"]] ?? 0.0;
    const I  = CIA_W[metrics["I"]] ?? 0.0;
    const A  = CIA_W[metrics["A"]] ?? 0.0;

    const f_impact = C === 0 && I === 0 && A === 0 ? 0 : 1.176;
    const impact   = 10.41 * (1 - (1 - C) * (1 - I) * (1 - A));
    const exploit  = 20 * AV * AC * Au;

    const base = ((0.6 * impact) + (0.4 * exploit) - 1.5) * f_impact;
    return Math.round(base * 10) / 10 + "";
  } catch {
    return null;
  }
}

// ── Helpers ───────────────────────────────────────────────────────────────────
function parseMetrics(vector) {
  const result = {};
  // Strip version prefix (CVSS:3.1/, CVSS:4.0/) or leave raw
  const body = vector.replace(/^CVSS:\d+\.\d+\//, "").replace(/^\(|\)$/g, "");
  for (const part of body.split("/")) {
    const [k, v] = part.split(":");
    if (k && v !== undefined) result[k] = v;
  }
  return result;
}

// CVSS v3 "roundup" — rounds to 1 decimal, always up
function roundUp(value) {
  const int = Math.round(value * 100000);
  if (int % 10000 === 0) return int / 100000;
  return (Math.floor(int / 10000) + 1) / 10;
}

// ── Main mutation helper (mirrors Python process_vulnerability) ───────────────
export function processVulnerability(vuln) {
  const severityList = vuln["severity"];

  if (Array.isArray(severityList) && severityList.length > 0) {
    const first = severityList[0];
    const vector = first["score"] || "";
    const type   = (first["type"] || "").toLowerCase();

    // Ubuntu label — no numeric score, pass string as-is
    if (type === "ubuntu_cve" || type === "ubuntu") {
      vuln["severity"]        = vector;
      vuln["severity_score"]  = null;
      vuln["severity_vector"] = null;
      return;
    }

    const { score, severity } = parse(vector);

    vuln["severity"]        = severity;
    vuln["severity_score"]  = parseFloat(score) || null;
    vuln["severity_vector"] = vector;

    // Fallback: if severity still unknown but we have a numeric score, bucket it
    if ((severity === "unknown" || severity === null) && score !== null) {
      vuln["severity"] = scoreToSeverity(score);
    }
    return;
  }

  vuln["severity"]        = "unknown";
  vuln["severity_score"]  = null;
  vuln["severity_vector"] = null;
}
